# torchbase
Python framework for microbial typing by reference

## Goals
Consistent reference format 
IPFS-based reference distribution
Modular allele detection?
Alignment parameter tuning in reference package
Parameter override in runtime
Typing:
	Antigen-based 
	wgMLST?
	Other stuff?
	2-stage approaches - alignment, then SNP calling
Stuff in the reference package:
	Allele references
	Allelic profile table
	Parameters for alignment and SNP calling
	Workflow description (stages, data flow)
Cool name
